ボウリングのスコア計算プログラムオブジェクト指向版の解答提出

yokomaru opened the pull request:
ボウリングのスコア計算プログラムオブジェクト指向版の課題を提出いたします。
お手隙でご確認お願いいたします！

# 課題URL
https://bootcamp.fjord.jp/products/20600

yokomaru committed:
first commit

yokomaru committed:
add Shot

- Shotクラスの追加
- Shotテストを追加

yokomaru committed:
Add Frame

- Frameクラスを追加
- Frameテストを追加

yokomaru committed:
Add Game

- Gameクラスを仮追加
- Gameのテストも仮追加

yokomaru committed:
Add Bowling

- main処理用のファイルを仮追加

yokomaru committed:
fix: separate game test and game file

yokomaru committed:
add: test_helper and test folder

- 記述が重複するため各テストファイルで読み込むminitestのライブラリをまとめた
- テストファイルをテストフォルダに移動

yokomaru committed:
fix: shot and test

- shotの点数に関する定数をGameからShotに移動
- 不要なreturnを三項演算子に変更
- rubocopでの指摘事項frozen_string_literal: trueを追加
- markは外部から参照される必要がないためattr_readerを削除し、それに伴いテストも削除

yokomaru committed:
fix: Frame and test

- Shotに移動した定数を参照する
- 外部から参照されない不要なattr_readerを削除し、インスタンス変数を直接参照するように修正
- 引数を必ず3つ渡すため、initializeのデフォルト値が不要になったので削除
- ストライクまたはスペアの分岐をGameからFrameに移動

yokomaru committed:
fix: Game

- shotsとscoreという変数名が区別つけづらかったので、score_textとscoreに変更
- スコアを配列にする処理を別メソッドに分ける(parse_score_text)
- スコアに関する定数をShotに移動する
- 実行処理を追加
- 使わなかったbowling.rbを削除
- rubocopでの指摘事項frozen_string_literal: trueを追加

yokomaru committed:
fix: delete comment in Game

- 削除し忘れたコメントを削除

yuuu commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,33 @@
+require_relative 'frame'
+
+class Game
+  GAME_TIMES = 10
```
この名前だと「ゲームの回数」「試合時間」のような意味に見えてしまうので、名前を改善したいです 🙏

yokomaru replied:
ご指摘ありがとうございます！
こちら、ゲームの回数ではなくフレームの回数のという意図だったため、`FRAME_TIMES`に修正いたします。

yuuu commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,33 @@
+  GAME_TIMES = 10
+
+  def initialize(score_text)
+    @score_text = score_text
```
今回はスコアを1回だけ計算するという要件なので引数のテキストを保持していても支障はないですが、例えば「同じゲームのスコアを複数回計算したい」「ゲームのスコア表の出力に対応したい」といった追加要件が発生すると、処理をするたびに `@score_text` のパースや `Frame.new` を行う必要が出てきます。

`initialize` の時点でそのあたりの処理を終えて、フレームの集合をインスタンス変数として保持する、というのはどうでしょうか？？

yokomaru replied:
コメントありがとうございます！自分の理解が足らず、2点質問させてください！🙏

> initialize の時点でそのあたりの処理を終えて、フレームの集合をインスタンス変数として保持する、というのはどうでしょうか？？

1. こちらでコメントいただいている**initialize の時点**とは以下どちらのタイミングのイメージでしょうかっ？
- initializeのメソッド内
- ARGVを受け取った時点(Game.newする前)

2. **フレームの集合をインスタンス変数として保持する**ということですが、受け取ったテキスト(ショットの集合データ)を配列にパースするという意味でしょうか？(`parse_score_text`を`score`処理より前に実行するイメージ)
それとも以下のように`score`計算処理の前にフレームごとに区切るところまで対応するようなイメージでしょうか？
```ruby
[[1,9,10], [10,1, 8], [1,8]...]
```

yuuu replied:
@yokomaru 
わかりにくくてすみません 🙏 

> こちらでコメントいただいているinitialize の時点とは以下どちらのタイミングのイメージでしょうかっ？

自分のコメント時点では `initializeのメソッド内` を想定していました。一方で、  `ARGVを受け取った時点(Game.newする前)` というのも戦略としてはアリです 👍 

> フレームの集合をインスタンス変数として保持するということですが、受け取ったテキスト(ショットの集合データ)を配列にパースするという意味でしょうか？(parse_score_textをscore処理より前に実行するイメージ)
> それとも以下のようにscore計算処理の前にフレームごとに区切るところまで対応するようなイメージでしょうか？

こちらは前者ですね。

今回はたまたまCLIアプリケーションとしてこのプログラムを書いていますが、例えば「Webアプリケーションに置き換えたい」といった要求が発生したときに、今の実装だと `Game` が持つインスタンス変数の内容が変わる可能性が高いです。クラスとはそのものの概念を表すものなので、「UIが変わっても本質部分の設計は変わらない」といえる設計にしたいです。

yuuu commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,33 @@
+  def score
+    scores = parse_score_text(@score_text)
+    total_point = 0
+    GAME_TIMES.times do
```
ここは `.sum` を使うと加算処理や `total_point` の宣言を自分で書かなくて良くなりますね。

```ruby
total_point = GAME_TIMES.times.sum do
  # 各フレームのスコアを戻り値にする
end
```

yokomaru replied:
`times`でEnumeratorが帰ってくるので`sum`が使用できるんですね・・！とても勉強になりました！
余計な処理を書く必要がなくなりとてもスッキリしました！

- [Integer\#times \(Ruby 3\.3 リファレンスマニュアル\)](https://docs.rubylang.org/ja/latest/method/Integer/i/times.html)
- [Enumerable\#sum \(Ruby 3\.3 リファレンスマニュアル\)](https://docs.rubylang.org/ja/latest/method/Enumerable/i/sum.html)

yuuu commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,19 @@
+
+  def score
+    if @first_shot.score == Shot::STRIKE_SCORE || @first_shot.score + @second_shot.score == Shot::SPARE_SCORE
+      [@first_shot.score, @second_shot.score, @third_shot.score].sum
```
複数のインスタンス変数に `.score` を付けて加算したいときは `map` を併用するとコードが少しだけスッキリします。

```ruby
[@first_shot, @second_shot, @third_shot].map(&:score).sum
```

yokomaru replied:
こちら修正いたします！

yuuu commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,33 @@
+    scores = parse_score_text(@score_text)
+    total_point = 0
+    GAME_TIMES.times do
+      frame = Frame.new(scores[0], scores[1], scores[2])
```
好みが分かれるとは思いますがこういった書き方もあります。こちらは修正必須ではないので参考まで 🙏 

```ruby
frame = Frame.new(*scores[0, 1], scores[2])
```

`scores[2]` は結果が `nil` になる可能性を考慮するので、まとめられない認識です。

yokomaru replied:
`*`で配列が展開されるんですね！
とても便利でスマートだと思ったので、
教えていただいた書き方で修正いたしました！

```ruby
frame = Frame.new(*scores[0, 2], scores[2]) #配列から要素を2つ取りたいので[0, 2]にしています
```

yokomaru committed:
fix: constant name GAME_TIMES to FRAME_TIMES

yokomaru committed:
fix: Game score method

- 計算結果保管用の変数とスコアの足し上げ処理を削除するためsumメソッドを使用する
- Frame.newの引数の渡し方を変更(*で配列を展開させる)

yokomaru committed:
fix: Frame score method

- shotインスタンスが都度scoreを読んでいたのでmap(&:score)に変更

yuuu commented on 06.bowling_object/shot.rb:
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class Shot
+  STRIKE_SCORE = 10
```
これら2つの定数は `Shot` ではなく `Frame` に存在すべきだと思いました。

ストライク・スペアというのはその1回の投球では断定ができず、2つの投球を組み合わせて初めてそれがストライクだったかスペアだったかがわかるためです。

`score` メソッドで `STRIKE_SCORE` を使っていますが、ここは「 `'X'` はストライク」というよりも「 `'X'` は `10`」として、ストライク華道家の判定は `Frame` に委ねるという考え方が良いです 👍

yokomaru committed:
fix: move full score constant Shot to Frame

- ストライクやスペアは単体のShotでは判定できずFrameで決まるため、Shotにある定数をFrameに移す
- ストライク・スペアの合計得点である10を入れる定数として命名をFULL_SCOREに変更
    - Shot内のscore内で単体のshotとFrameのFULL_SCOREで比較している箇所を単純に満点の10点と比較するように修正
- shotの点数を定数と比較して分岐していた箇所が長くなりすぎたため、strike?とspare?メソッドを作成

yokomaru committed:
fix: change  instance variable value  String to Array  in Game

- Game内のインスタンス変数にターミナルから送られてくるテキストをそのまま保持しているが、今後再利用する際都度配列へのパースが必要になるため、initialize内で配列に変換する
- 配列変換時にFrameに変換することでscore計算の処理を簡潔にする
- Frameに定義したstrike?メソッドの使用

yokomaru committed:
add: Bowling

- Gameクラス内に実行処理を書いていたが、拡張性を考慮(複数ゲームを行いたい時など)しGameを実行する処理を追加
- Gameを渡すことで依存性を注入する

yokomaru committed:
fix: Shot have point as a variable

- FrameからShotを呼ぶ際、毎回scoreメソッドで文字列を数値にパースしていたので、@scoreという変数にデータとして数値の情報を持っておくように修正
- 文字列はmarkとしてそのまま持っておく

yokomaru committed:
add: Frame have shots

- Shotを複数計算する処理を複数箇所で行うため、インスタンス変数にArrayのshotsを持たせる
- 計算処理を行う際shotsを用いる

yokomaru commented on 06.bowling_object/shot.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def initialize(mark)
+    @mark = mark
+    @score = parse_score
```
渡されてくる文字列は`@mark`として持ちつつ、intに変換した`@score`も持たせるように修正

yokomaru commented on 06.bowling_object/frame.rb:
```diff
@@ -0,0 +1,30 @@
+    @first_shot = Shot.new(first_mark)
+    @second_shot = Shot.new(second_mark)
+    @third_shot = Shot.new(third_mark)
+    @shots = [@first_shot, @second_shot, @third_shot]
```
`score`や `spare?`でshotを配列として扱うため、`@shots`として配列にしたデータを持つようにした

yuuu replied:
いいと思います 👍 `Frame#strike?` 内の `@first_shot` を `@shots.first` に置き換えれば、 `@first_shot` `@second_shot` `@third_shot` は削除できそうですね。

yokomaru replied:
ありがとうございます！
`.first`でシンプルな記述で先頭を取得できました！

yokomaru committed:
delete: score check

- parse_score_textの時点でintに変換する必要はないため変換処理を削除

yokomaru commented on 06.bowling_object/game.rb:
```diff
@@ -0,0 +1,26 @@
+    scores = score_text.split(',')
+    Array.new(FRAME_TIMES) do
+      frame = Frame.new(*scores[0, 2], scores[2])
+      scores.shift(frame.strike? ? 1 : 2)
```
ペアプロ時にお話があったshiftを他の処理に書き換える方法を検討しましたが、
良い案が浮かばなかったのでこのままにしてあります。

yokomaru commented on 06.bowling_object/bowling.rb:
```diff
@@ -0,0 +1,18 @@
+
+require_relative 'game'
+
+class Bowling
```
gameを実行するBowlingクラスを追加。
BowlingクラスでFrameの配列にすることも検討したが、
Gameの上位存在であるBowlingがFrameを知っていることに違和感を感じたためあくまで実行するだけのクラスとしている

yokomaru committed:
fix: arguments change string to array

- 引数は配列のshotsのみ変更する
    - @first_shot、@second_shot、@third_shotを削除
    - 引数を配列として受け取る
    - strike?の処理は配列の先頭(first_shot)を取得する

yokomaru committed:
fix: method name change parse_score_text to generate_frames

- 文字列をパースするというより、frameの配列を作成する処理に変更になったので、メソッド名をgenerate_framesに変更
